### DISCUSSION TITLE
how to use a di injected custom rabbitmq client

### DISCUSSION THREAD

**Discussion by vigouredelaruse** [Azure/azure-functions-durable-extension#2448] [Q&A] [answered]
the following scenario causes a rain of memory leaking rabbitmq connection exceptions
- a custom rabbitmq client instantiated by orchestrator implementation class constructor dependency injection
- c# in-process with startup.cas functionsstartup 
- builder.Services.AddScoped<IHorselessRabbitMQClient, HorselessRabbitMQClient>();
- rabbitmq client binds to configuration and opens connection (but not channel) in constructor

the use case calls for 
- a rabbitmq client that dynamically addresses an arbitrary set of exchanges and queues (outbound)
- rabbit mq client invoked from [ActivityTrigger] or  [DurableClient] equpped  orchestration starter function

please advise on the viability of the scenario and alternate non-memory leaking approaches that satisfy the dynamic rabbitmq connection constraint at minimum

**Comment by cgillum** [comment, marked as answer]
I'm unfortunately not familiar enough with dependency injection and how it works with scoped services. I'm also not familiar enough with the RabbitMQ client library you're using to know whether it plays nicely with scoped services - for example, when does the connection get closed?

I don't think this problem is necessarily specific to Durable Functions, but you might have better luck using a singleton connection factory and managing the lifetime of connections (e.g. connection pooling) yourself.

**Reply by vigouredelaruse** [reply]
took a look at "how masstransit does it" and this (https://groups.google.com/g/masstransit-discuss/c/j7_24OzZmXQ?pli=1) discussion strongly suggests a tendency towards singleton service with a 'permanently' open connection
